lib/api/src/entities/function/host/mod.rs

```diff
@@ -0,0 +1,103 @@
+mod imp;
+
+use crate::{
+    vm::{VMFunctionCallback, VMTrampoline},
+    WasmTypeList,
+};
+
+/// The `HostFunction` trait represents the set of functions that
+/// can be used as host function. To uphold this statement, it is
+/// necessary for a function to be transformed into a
+/// `VMFunctionCallback`.
+pub trait HostFunction<T, Args, Rets, Kind>
+where
+    Args: WasmTypeList,
+    Rets: WasmTypeList,
+    Kind: HostFunctionKind,
+{
+    #[cfg(feature = "jsc")]
+    /// Get the pointer to the function body for a given runtime.
+    fn jsc_function_callback(&self) -> crate::rt::jsc::vm::VMFunctionCallback;
```

Comment:
Having backend-specific functions in a trait, that are enabled conditionally, seems like a huge hack.

Why is this needed?

Reply:
I agree, it's hacky and ugly. The problem is [this](https://github.com/wasmerio/wasmer/blob/5198-cli-support-wasm_c_api-backends-together-with-native-ones/lib/api/src/entities/function/mod.rs#L112) function - and the sister `_with_env` function - which has a constraint on the parameter `F` being a `HostFunction<...>`.

We could address this with per-runtime traits (e.g. `SysHostFunction`, `WamrHostFunction` and so on), but new problems would arise, in particular: what would be the signature of the top-level `Function::new_typed(_with_env)` function? In some way, we'll need to be able to make the `F` parameter into a `SysHostFunction` when actually calling the internal API to create the `Function` (`Sys::new_typed`, so to speak). If we keep it `HostFunction`: 
```
pub trait HostFunction<T, Args, Rets, Kind>
where
    Args: WasmTypeList,
    Rets: WasmTypeList,
    Kind: HostFunctionKind,
{
   /* empty */
}

pub trait SysHostFunction<T, Args, Rets, Kind>
where
    Args: WasmTypeList,
    Rets: WasmTypeList,
    Kind: HostFunctionKind,
{
   fn fn_callback(&self) -> ...;
}
....

/// Creates a new host `Function` from a native function.
pub fn new_typed<F, Args, Rets>(store: &mut impl AsStoreMut, func: F) -> Self
where
        F: HostFunction<(), Args, Rets, WithoutEnv> + 'static + Send + Sync,
        Args: WasmTypeList,
        Rets: WasmTypeList,
{ 
      ...
      cfg sys  {
         ...sys::Function::new_typed(store, f /* <cast to SysHostFunction if not automatic>/*)
      }
} 

// in sys
/// Creates a new host `Function` from a native function.
pub(crate) fn new_typed<F, Args, Rets>(store: &mut impl AsStoreMut, func: F) -> Self
where
        F: SysHostFunction<(), Args, Rets, WithoutEnv> + 'static + Send + Sync,
        Args: WasmTypeList,
        Rets: WasmTypeList,
{
    let cback = func.fn_callback();
    ....
}
```
We then have to implement something like 
```
impl< $( $x, )* Rets, RetsAsResult, Func> SysHostFunction<(), ( $( $x ),* ), Rets, WithoutEnv> for Func where
    $( $x: FromToNativeWasmType, )*
    Rets: WasmTypeList,
    RetsAsResult: IntoResult<Rets>,
    T: Send + 'static,
    Func: HostFunction<(), $($x),*), Rets, WithoutEnv> + 'static,
{
   fn fn_callback(&self) -> ...  {
       /* How do we keep track of Func being a callable object here? */ 
   }  
}
```
A possibility would be to make `HostFunction: Fn(Args) -> R` where `R: IntoResult<Rets>`, but I'm not sure it would work. What do you think? Probably, a fresh pair of eyes can see a completely different way to do it :)

Reply:
[81f4195](https://github.com/wasmerio/wasmer/pull/5243/commits/81f41957f94a32924733020ad2373d0158fe60d1)